darshan-util/pydarshan/benchmarks/benchmarks/dxt_heatmap.py

```diff
@@ -69,8 +69,8 @@ def setup(self, unique_ranks, bin_count, density):
     def time_get_heatmap_df(self, unique_ranks, bin_count, density):
         # benchmark get_heatmap_df() handling of variable
         # numbers of unique ranks/bins
-        heatmap_handling.get_heatmap_df(self.agg_df, xbins=bin_count, nprocs=unique_ranks)
+        heatmap_handling.get_heatmap_df(self.agg_df, xbins=bin_count, nprocs=unique_ranks, tmax=1)
 
 
     def peakmem_get_heatmap_df(self, unique_ranks, bin_count, density):
-        heatmap_handling.get_heatmap_df(self.agg_df, xbins=bin_count, nprocs=unique_ranks)
+        heatmap_handling.get_heatmap_df(self.agg_df, xbins=bin_count, nprocs=unique_ranks, tmax=1)
```

Comment:
same comment on the "magic number" for time here--this feels a bit fragile IMO, instead of being extracted programmatically from a report for example.

Reply:
Alright, I'll can change this to use something like `agg_df["end_time"].max()`.

Reply:
Hmm, maybe, I'm kind of thinking this should just happen inside `get_hetmap_df()` or something. The design my warrant some careful consideration.

If `get_heatmap_df()` is going to be used for generating runtime maps as well, it could just use all the timing information available to find an appropriate upper timing bound internally. It depends on where the decision about which heatmaps to plot is going to be made I suspect.

Reply:
If the decision is only relevant in the plotting code, which it may be (?), then a function to inspect the contents of `agg_df` and `runtime_hmap_df` (when available) or whatever might make sense for setting `max_t`.

Then the plotting code retrieves the DXT/runtime datastructures with their "common denomintaor" `max_t`? I guess then I wonder what use we'd have for padding, since the data would always be bound to the observed upper limit. Can runtime heatmap final bin ever finish before the DXT max end time?

Maybe the Argonne folks have a different vision though.

Reply:
Yeah I think I would like to handle the heatmap retrieval (either DXT or runtime) in `get_heatmap_df()`. There is a challenge there because the heatmaps are stored under generic "POSIX" and "MPIIO" module names, but I think if we pass along the report object to `get_heatmap_df()` we can do something like the following:

```python
# retrieve DXT POSIX heatmap
hmap_df = get_heatmap_df(report, mod="DXT_POSIX", ops=["read", "write"])
# retrieve runtime POSIX heatmap
hmap_df = get_heatmap_df(report, mod="POSIX", ops=["read", "write"])
```

Passing the report object would remove the need for `nprocs` or `tmax` as formal parameters, and we can handle cases based on whether the `HEATMAP` module is present or not. We would have to move the call to `get_aggregate_data()` inside of `get_heatmap_df()` as well, but I think I like that better anyways and this makes some of the `tmax` cases [here](https://github.com/darshan-hpc/darshan/pull/696#issuecomment-1071583214) pretty easy I think.

Reply:
After thinking about this more, I think it might be nice to have the following functions defined:
1. `get_dxt_heatmap_df(report, mod, ops, xbins, tmax=None)`: basically the current `get_heatmap_df()` function renamed, with the call to `get_aggregate_data()` moved inside the function body (it is only needed to get the DXT heatmap)
2. `get_runtime_heatmap_df(report, mod, ops, tmax=None)`: new function for handling the runtime heatmap retrieval, should be pretty simple. Maybe so simple it is unnecessary
3. `get_heatmap_tmax(report, mod)`: determines what the maximum x-axis value should be for a heatmap depending on the module input
4. `get_heatmap_df(report, mod, ops, xbins=None)`: function for handling all of the above function calls. This would be the function you would call to actually retrieve the dataframe for a given module.

With a default `tmax=None` for both DXT/runtime heatmap retrieval functions they would use their "normal" methods for calculating `tmax`, but we would have the option to override it when we wanted to (i.e. for multiple heatmap cases). For plotting in `plot_heatmap()` we can retrieve `tmax` from the column names/intervals in the returned dataframe from `get_heatmap_df()`.

This could be accomplished in a series of smaller PR's:
1.  move `get_aggregate_data()` call into `get_heatmap_df()` body, fix x-ticks functions to receive `tmax` as input (to remove dependence on `agg_df()`)
2. rename `get_heatmap_df()` to `get_dxt_heatmap_df()`
3. add `get_runtime_heatmap_df()` (maybe with a unit test)
4. add `get_heatmap_tmax()` and `get_heatmap_df()`, and generalize `plot_heatmap()` so it can plot both heatmap types. This is where we would decide how to handle setting `tmax` for the various cases and actually fix gh-575

Does this seem sensible? I would probably create a new PR and close this one since most of the changes here are likely unnecessary.  But I could handle 1., 3., and parts of 4. at the same time because they wouldn't overlap I don't think.

Reply:
Hard to say, sounds like a fair bit of refactoring.

I thought `plot_heatmap()` would basically just need to retrieve some form of raw data for the runtime and DXT heatmaps to find a common `tmax`, then retrieve dataframes adjusted to that `tmax`.

You could check with the team this afternoon to make sure you're absolutely clear on the summary report requirements for the different heatmap permutation scenarios.